connector/src/main/java/com/google/cloud/hive/bigquery/connector/output/direct/DirectWriterContext.java

```diff
@@ -0,0 +1,144 @@
+   * @param bigQuerySchema the bigQuery schema
+   * @return The TableId to which Hive will do the writing: whether that is the destinationTableID
+   *     or the temporaryTableId.
+   */
+  private BigQueryTable getOrCreateTable(
+      TableId destinationTableId, com.google.cloud.bigquery.Schema bigQuerySchema)
+      throws IllegalArgumentException {
+    if (bigQueryClient.tableExists(destinationTableId)) {
+      TableInfo destinationTable = bigQueryClient.getTable(destinationTableId);
+      com.google.cloud.bigquery.Schema tableSchema = destinationTable.getDefinition().getSchema();
+      Preconditions.checkArgument(
+          BigQueryUtil.schemaEquals(tableSchema, bigQuerySchema, /* regardFieldOrder */ false),
+          new BigQueryConnectorException.InvalidSchemaException(
+              "Destination table's schema is not compatible with query's" + " schema"));
+      return new BigQueryTable(destinationTable.getTableId(), false);
+    } else {
+      return new BigQueryTable(
+          bigQueryClient.createTable(destinationTableId, bigQuerySchema).getTableId(), true);
+    }
+  }
+
+  public void commit(List<String> streamNames) {
+    if (writingMode.equals(WritingMode.IGNORE_INPUTS)) return;
+    BatchCommitWriteStreamsRequest.Builder batchCommitWriteStreamsRequest =
+        BatchCommitWriteStreamsRequest.newBuilder().setParent(tablePathForBigQueryStorage);
+    for (String streamName : streamNames) {
+      batchCommitWriteStreamsRequest.addWriteStreams(streamName);
+    }
+    BatchCommitWriteStreamsResponse batchCommitWriteStreamsResponse =
+        writeClient.batchCommitWriteStreams(batchCommitWriteStreamsRequest.build());
+
+    if (!batchCommitWriteStreamsResponse.hasCommitTime()) {
+      throw new BigQueryConnectorException(
+          "BigQuery writer failed to batch commit its BigQuery write-streams");
+    }
+
+    LOG.info(
+        "BigQuery writer has committed at time: {}",
+        batchCommitWriteStreamsResponse.getCommitTime());
+
+    if (writingMode.equals(WritingMode.OVERWRITE)) {
+      Job overwriteJob =
+          bigQueryClient.overwriteDestinationWithTemporary(
+              tableToWrite.getTableId(), destinationTableId);
+      BigQueryClient.waitForJob(overwriteJob);
+      Preconditions.checkState(
+          bigQueryClient.deleteTable(tableToWrite.getTableId()),
+          new BigQueryConnectorException(
+              String.format("Could not delete temporary table %s from BigQuery", tableToWrite)));
+    }
+  }
+
+  // Used for the getOrCreateTable output, to indicate if the table should be deleted on abort
+  static class BigQueryTable {
+    private final TableId tableId;
+    private final boolean toDeleteOnAbort;
+
+    public BigQueryTable(TableId tableId, boolean toDeleteOnAbort) {
+      this.tableId = tableId;
+      this.toDeleteOnAbort = toDeleteOnAbort;
+    }
+
+    public TableId getTableId() {
+      return tableId;
+    }
+
+    public boolean toDeleteOnAbort() {
```

Comment:
Are we using the `toDeleteOnAbort` variable anywhere? I don't see any usages

Reply:
Good catch. I've added the logic to use it.